lnwire/onion_message.go

```diff
@@ -0,0 +1,79 @@
+package lnwire
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+)
+
+// OnionMessage is a message that carries an onion-encrypted payload.
+// This is used for BOLT12 messages.
+type OnionMessage struct {
+	// BlindingPoint is the route blinding ephemeral pubkey to be used for
+	// the onion message.
+	BlindingPoint *btcec.PublicKey
+
+	// OnionBlob is the raw serialized mix header used to relay messages in
+	// a privacy-preserving manner. This blob should be handled in the same
+	// manner as onions used to route HTLCs, with the exception that it uses
+	// blinded routes by default.
+	OnionBlob []byte
+}
+
+// NewOnionMessage creates a new OnionMessage.
+func NewOnionMessage(blindingPoint *btcec.PublicKey,
+	onion []byte) *OnionMessage {
+
+	return &OnionMessage{
+		BlindingPoint: blindingPoint,
+		OnionBlob:     onion,
+	}
+}
+
+// A compile-time check to ensure OnionMessage implements the Message interface.
+var _ Message = (*OnionMessage)(nil)
+
+// Decode reads the bytes stream and converts it to the object.
+func (o *OnionMessage) Decode(r io.Reader, pver uint32) error {
+	if err := ReadElement(r, &o.BlindingPoint); err != nil {
+		return err
+	}
+
+	var onionLen uint16
+	if err := ReadElement(r, &onionLen); err != nil {
+		return err
+	}
+
+	o.OnionBlob = make([]byte, onionLen)
+	if err := ReadElement(r, o.OnionBlob); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Encode converts object to the bytes stream and write it into the
+// write buffer.
+func (o *OnionMessage) Encode(w *bytes.Buffer, pver uint32) error {
+	if err := WriteElement(w, o.BlindingPoint); err != nil {
+		return err
+	}
+
+	onionLen := len(o.OnionBlob)
+	if err := WriteElement(w, uint16(onionLen)); err != nil {
```

Comment:
Why not use varint ?

Reply:
I think `WriteUint16` is maybe even a better choice. `WriteElement` is about to be decommissioned so I changed the other references as well.